openpgp/packet/aead_encrypted.go

```diff
@@ -0,0 +1,408 @@
+// Copyright (C) 2019 ProtonTech AG
+
+package packet
+
+import (
+	"bytes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+
+	"golang.org/x/crypto/eax"
+	"golang.org/x/crypto/ocb"
+	"golang.org/x/crypto/openpgp/errors"
+	"golang.org/x/crypto/openpgp/internal/algorithm"
+)
+
+// AEADEncrypted represents an AEAD Encrypted Packet (tag 20, RFC4880bis-5.16).
+type AEADEncrypted struct {
+	prefix       []byte    // Packet tag, version, cipher, mode, chunk size byte
+	initialNonce []byte    // Referred to as IV in RFC4880-bis
+	Contents     io.Reader // Encrypted chunks and tags
+}
+
+// An AEAD opener/sealer, its configuration, and data for en/decryption.
+type worker struct {
+	aead         cipher.AEAD
+	config       *AEADConfig
+	header       []byte // Chunk-independent associated data
+	index        []byte // Chunk counter
+	initialNonce []byte
+	cache        []byte
+}
+
+// streamWriter encrypts and writes bytes. It encrypts when necessary according
+// to the AEAD block size, and caches the extra encrypted bytes for next write.
+type streamWriter struct {
+	worker                               // Embedded plaintext sealer
+	writer                io.WriteCloser // 'writer' is a partialLengthWriter
+	writtenEncryptedBytes int
+}
+
+// streamReader reads and decrypts bytes. It caches extra decrypted bytes when
+// necessary, similar to streamWriter.
+type streamReader struct {
+	worker                       // Embedded ciphertext opener
+	reader             io.Reader // 'reader' is a partialLengthReader
+	carry              []byte    // Used to detect last chunk
+	readPlaintextBytes int
+}
+
+func (ae *AEADEncrypted) parse(buf io.Reader) error {
+	ptype, _, contentsReader, err := readHeader(buf)
+	if ptype != packetTypeAEADEncrypted || err != nil {
+		return errors.AEADError("Error reading packet header")
+	}
```

Comment:
I think in the `packet` interface, the packet header will have already been read at this point, since otherwise `packet.Read` wouldn't know which packet to construct.